lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
+    auto AllowedRange = llvm::enumerate(llvm::reverse(Allowed));
+    for (auto Pair : AllowedRange) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
+      model::Register::Values Register = Pair.value();
+      auto UsedIterator = Used.find(Register);
+      if (UsedIterator != Used.end())
+        ResultingTypes.emplace_back(UsedIterator->Type);
+      else if (MustUseTheNextOne)
+        ResultingTypes.emplace_back(std::nullopt);
+
+      MustUseTheNextOne = MustUseTheNextOne || (UsedIterator != Used.end());
+    }
+
+    revng_assert(ResultingTypes.size() != 0);
+    if (ResultingTypes.size() == 1) {
+      // Short-circuit for a single register return value.
+      auto Kind = model::Register::primitiveKind(Allowed.at(0));
+      if (ResultingTypes.front() != std::nullopt) {
+        return ensureTypeIsAvailable(*ResultingTypes.front(),
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      } else {
+        // No information about the type of this register, use the default.
+        return ensureTypeIsAvailable(model::QualifiedType{},
+                                     Allowed.at(0),
+                                     Binary,
+                                     TypeDependencies);
+      }
+    } else if (ResultingTypes.size() == 2) {
+      // Short-circuit for a double register return value.
```

Comment:
It seems to me that if we're returning two pointers, instead of a `struct { A *First; B *Second }` we'll get a `uint128_t`.
Maybe you wanted to add something like "and both types are primitives of the same kind"?

Reply:
We discussed that later and for now `uint128_t` initiative is totally abandoned and we'll always get the struct, especially considering that updated `Layout`s now also support that better (and revng-c side also expects with my WIP patch).

Reply:
> You mean this code is not up-to-date?

This is one of pieces I'm actively working on based on the feedback from Pietro and the discoveries I made when adapting the revng-c to use this.
(having an easy way to compare `dwarf` vs `result of our conversion` side to side on a bunch of libc or whatever functions opened my eyes to some things I kind of missed).

Now I'm REALLY-REALLY unsure we actually want to try and merge two 64-bit `int`s into a 128-bit one. Because there are a lot of cases where `std::pair<Pointer *, AnotherPointer *>` is being returned. And since from our point of view that's just a normal `rdx-rax` return value, they get swept as `uint128_t` which is probably not what we want. Like at all.

Reply:
I agree, merging is an advanced thing to do. We should support the basic situation (i.e., separate them), which works but it's ugly if they were actually a `uint128_t`, and then, at the next QA iteration, decide how to take on these issues.

Our current scope is to support stack arguments and in "regular situations".

Reply:
Pretty much, yeah.

Reply:
As sad as it make me, this is done.